src/test/fuzz/script_sign.cpp

```diff
@@ -0,0 +1,149 @@
+    if (k.IsValid()) {
+        provider.AddKey(k);
+    }
+
+    {
+        const Optional<CMutableTransaction> mutable_transaction = ConsumeDeserializable<CMutableTransaction>(fuzzed_data_provider);
+        const Optional<CTxOut> tx_out = ConsumeDeserializable<CTxOut>(fuzzed_data_provider);
+        const unsigned int n_in = fuzzed_data_provider.ConsumeIntegral<unsigned int>();
+        if (mutable_transaction && tx_out && mutable_transaction->vin.size() > n_in) {
+            SignatureData signature_data_1 = DataFromTransaction(*mutable_transaction, n_in, *tx_out);
+            CTxIn input;
+            UpdateInput(input, signature_data_1);
+            const CScript script = ConsumeScript(fuzzed_data_provider);
+            SignatureData signature_data_2{script};
+            signature_data_1.MergeSignatureData(signature_data_2);
+        }
+        if (mutable_transaction) {
+            CTransaction tx_from{*mutable_transaction};
+            CMutableTransaction tx_to;
+            const Optional<CMutableTransaction> opt_tx_to = ConsumeDeserializable<CMutableTransaction>(fuzzed_data_provider);
+            if (opt_tx_to) {
+                tx_to = *opt_tx_to;
+            }
+            CMutableTransaction script_tx_to = tx_to;
+            CMutableTransaction sign_transaction_tx_to = tx_to;
+            if (n_in < tx_to.vin.size() && tx_to.vin[n_in].prevout.n < tx_from.vout.size()) {
+                (void)SignSignature(provider, tx_from, tx_to, n_in, fuzzed_data_provider.ConsumeIntegral<int>());
+            }
+            if (n_in < script_tx_to.vin.size()) {
+                (void)SignSignature(provider, ConsumeScript(fuzzed_data_provider), script_tx_to, n_in, ConsumeMoney(fuzzed_data_provider), fuzzed_data_provider.ConsumeIntegral<int>());
+                MutableTransactionSignatureCreator signature_creator{&tx_to, n_in, ConsumeMoney(fuzzed_data_provider), fuzzed_data_provider.ConsumeIntegral<int>()};
+                std::vector<unsigned char> vch_sig;
+                CKeyID address;
+                if (fuzzed_data_provider.ConsumeBool()) {
+                    if (k.IsValid()) {
+                        address = k.GetPubKey().GetID();
+                    }
+                } else {
+                    address = CKeyID{ConsumeUInt160(fuzzed_data_provider)};
+                }
+                (void)signature_creator.CreateSig(provider, vch_sig, address, ConsumeScript(fuzzed_data_provider), fuzzed_data_provider.PickValueInArray({SigVersion::BASE, SigVersion::WITNESS_V0}));
+            }
+            std::map<COutPoint, Coin> coins;
+            while (fuzzed_data_provider.ConsumeBool()) {
+                const Optional<COutPoint> outpoint = ConsumeDeserializable<COutPoint>(fuzzed_data_provider);
+                if (!outpoint) {
+                    break;
+                }
+                const Optional<Coin> coin = ConsumeDeserializable<Coin>(fuzzed_data_provider);
+                if (!coin) {
+                    break;
```

Comment:
unrelated, but I was wondering if `ConsumeDeserializable<T>` could always return type `T` (if the stream was empty, it would return `T{}`). Not returning optional would bring this helper in line with all other helpers. For example `ConsumeRandomLengthString` always return `std::string`.

Reply:
You mean something along the lines of:

```c++
template <typename T>
NODISCARD inline T ConsumeDeserializable(FuzzedDataProvider& fuzzed_data_provider, const size_t max_length = 4096) noexcept
{
    const std::vector<uint8_t> buffer = ConsumeRandomLengthByteVector(fuzzed_data_provider, max_length);
    CDataStream ds{buffer, SER_NETWORK, INIT_PROTO_VERSION};
    T obj;
    try {
        ds >> obj;
    } catch (const std::ios_base::failure&) {
    }
    return obj;
}
```

I've thought about that and while it would be better from a developer ergonomics perspective I think it might come with some negative impact on overall fuzzing. One risk I see is that the fuzzer might try to work also on "meaningless" inputs (inputs which cause `std::ios_base::failure`) which would now become "somewhat meaningful" due to default construction.

Take the extreme example of a huge input containing a few million bytes worth of 0x41 scream (`AAAAAAAAAAA…`) being deserialized to an object of type `CFoo`. The current fuzzer would quickly reach `std::ios_base::failure`, return a `nullopt` and then give up without further processing (fail early thanks to `if (!opt_foo) { return; }` or similar). If we were to return a `CFoo{}` instead then the fuzzer would proceed with processing (fail late) which may slow down the fuzzing.

Does it make sense? :)

tl;dr -- would be worth doing if no negative fuzzing speed impact can be measured :)

Reply:
I think when it comes to trading off giving the fuzz test more freedom to explore more code branches vs micro-optimizing for performance, we should always pick the option that gives the fuzz engine more freedom. Fuzzing is great to complement traditional tests (e.g. our unit and functional tests). Those traditional tests offer a reasonable line coverage, but fuzz testing can offer great branch coverage. Giving the fuzz engine to explore as much branches as possible is vital. With modern coverage guided fuzz engines, the strength of fuzzing is that even with a large search space, the fuzz engine will always find useful inputs that increase coverage. Additional search space may (or may not!) come with a performance penalty, but I think we shouldn't use performance as a way to measure usefulness of a fuzz tests. The fastest fuzz test is the one that tests nothing. However, it is also the most useless one. Anything that can be solved by letting the fuzz engine run for another hour or by simply adding one more core should not be a reason to cripple the test into one that yields less coverage.

In fact, this has shown to be effective in the past. Take for example the process_message harness vs the process_message*s* harness. The latter is allowed to send an arbitrary number of messages and start an arbitrary number of peers. The latter (more flexible and broader) test that gives the engine more freedom is also the *only* test that has found bugs in pull requests that none of the other fuzz test or unit test or even functional tests have found. Specifically:

* It found a nullptr dereference: https://github.com/bitcoin/bitcoin/pull/18808#discussion_r417307258
* It found an unidentified problem: https://github.com/bitcoin/bitcoin/pull/15197#pullrequestreview-408731734
* Some more that I am missing...?

Reply:
I agree regarding the coverage vs performance trade-off: I'd choose "explore more code branches" over performance any day in week :)

Do you see a scenario where returning a default constructed object in case of serialisation failure is likely to give more freedom to explore more code branches compared to not doing so?

If so we'd have _two_ reasons to do this: developer ergonomics (no more `std::optional`) _and_ increased coverage :)

My thinking is more along the lines of "how do we guide the coverage-guided fuzzer in a meaningful way" rather than "micro-optimizing for performance". I guess the question boils down to if and to what extent the guidance provided by failing early is meaningful for the fuzzer :)

Code if anyone wants to experiment: https://github.com/bitcoin/bitcoin/compare/master...practicalswift:consume

Reply:
> Do you see a scenario where returning a default constructed object in case of serialisation failure is likely to give more freedom to explore more code branches compared to not doing so?

Unclear. This question boils down to whether default initialization leaves members uninitialized or initializes them with values that can not be the result of deserialization.

Reply:
Agreed. Sounds like we agree on the need for experimentation and measuring before proceeding with this :)

Reply:
@practicalswift Are you going to create a pull with the branch or should I do it?

Reply:
@MarcoFalke I plan to do a FuzzBench style test between the two variations (`T ConsumeDeserializable(…)` and `std::optional<T> ConsumeDeserializable(…)`) to see which one that helps us reach the highest coverage after 24 hours of fuzzing using `libFuzzer` (with each harness given `1/N` slice of the runtime) with an empty input corpus. Does that sound like a fair experimental setup that could guide us regarding the trade-offs we're facing here? If so, I'll do it :)

Reply:
Sure, can't hurt to do a benchmark